In [67]:
import pandas as pd

df = pd.read_csv("weather_data_1hr.csv")

In [68]:
!pip3 install torch

ERROR: Could not find a version that satisfies the requirement torch (from versions: none)
ERROR: No matching distribution found for torch


In [33]:
df

loc_id        date  time isdaytime  tempC  tempF  windspeedMiles  \
0          1  2019-06-01     0        no     16     61               6   
1          1  2019-06-01   100        no     16     60               6   
2          1  2019-06-01   200        no     15     59               5   
3          1  2019-06-01   300        no     15     58               5   
4          1  2019-06-01   400        no     15     59               4   
...      ...         ...   ...       ...    ...    ...             ...   
1459       1  2019-07-31  1900        no     20     68              11   
1460       1  2019-07-31  2000        no     20     67              11   
1461       1  2019-07-31  2100        no     19     66              10   
1462       1  2019-07-31  2200        no     18     65               9   
1463       1  2019-07-31  2300        no     18     64               8   

      windspeedKmph  winddirdegree winddir16point  ...  HeatIndexC HeatIndexF  \
0                10            228             SW  ...          16         61   
1                 9            230             SW  ...          16         60   
2                 9            233             SW  ...          15         59   
3                 8            235             SW  ...          15         58   
4                 7            229             SW  ...          15         59   
...             ...            ...            ...  ...         ...        ...   
1459             18            260              W  ...          20         68   
1460             17            258            WSW  ...          20         67   
1461             15            257            WSW  ...          19         66   
1462             15            258            WSW  ...          18         65   
1463             14            259            WSW  ...          18         64   

     DewPointC  DewPointF  WindChillC  WindChillF  WindGustMiles  \
0           12         54          16          61              7   
1           12         53          16          60              7   
2           11         53          15          59              6   
3           11         52          14          58              6   
4           11         52          15          59              5   
...        ...        ...         ...         ...            ...   
1459         8         47          20          68             14   
1460         9         48          20          67             14   
1461        10         49          19          66             13   
1462        10         50          18          65             12   
1463        11         52          18          64             11   

      WindGustKmph  FeelsLikeC  FeelsLikeF  
0               12          16          61  
1               11          16          60  
2               10          15          59  
3                9          14          58  
4                8          15          59  
...            ...         ...         ...  
1459            23          20          68  
1460            22          20          67  
1461            22          19          66  
1462            20          18          65  
1463            18          18          64  

[1464 rows x 31 columns]

In [69]:
df.columns

Index(['loc_id', 'date', 'time', 'isdaytime', 'tempC', 'tempF',
       'windspeedMiles', 'windspeedKmph', 'winddirdegree', 'winddir16point',
       'weatherCode', 'weatherIconUrl', 'weatherDesc', 'precipMM',
       'precipInches', 'humidity', 'visibilityKm', 'visibilityMiles',
       'pressureMB', 'pressureInches', 'cloudcover', 'HeatIndexC',
       'HeatIndexF', 'DewPointC', 'DewPointF', 'WindChillC', 'WindChillF',
       'WindGustMiles', 'WindGustKmph', 'FeelsLikeC', 'FeelsLikeF'],
      dtype='object')

In [35]:
df = df.drop(["loc_id","time","date"], axis=1)


In [36]:
def is_numeric(col):
    try:
      # df_min_max_scaled = df_min_max_scaled.drop(["date", "time"], axis=1)
      pd.to_numeric(col)
      return True
    except:
        return False

In [37]:
#Find all the numerical columns
numeric_columns = []
for column in df.columns:
  if(is_numeric(df[column])):
    numeric_columns.append(column)

In [38]:
print(numeric_columns)

['tempC', 'tempF', 'windspeedMiles', 'windspeedKmph', 'winddirdegree', 'weatherCode', 'precipMM', 'precipInches', 'humidity', 'visibilityKm', 'visibilityMiles', 'pressureMB', 'pressureInches', 'cloudcover', 'HeatIndexC', 'HeatIndexF', 'DewPointC', 'DewPointF', 'WindChillC', 'WindChillF', 'WindGustMiles', 'WindGustKmph', 'FeelsLikeC', 'FeelsLikeF']


In [39]:
#normalization of nuerical data i.e between 0 and 1
#MinMax Scaling for all umeric colums
#MinMAx scaling allows us to get information about the significance of each entry in a column
df_min_max_scaled = df.copy()
for numeric_column in numeric_columns:
  df_min_max_scaled[numeric_column] = (df_min_max_scaled[numeric_column] - df_min_max_scaled[numeric_column].min()) / (df_min_max_scaled[numeric_column].max() - df_min_max_scaled[numeric_column].min())

In [40]:
df_min_max_scaled.head()
# We do ot need the location id at this point

isdaytime     tempC     tempF  windspeedMiles  windspeedKmph  winddirdegree  \
0        no  0.366667  0.370370        0.285714        0.28125       0.640227   
1        no  0.366667  0.351852        0.285714        0.25000       0.645892   
2        no  0.333333  0.333333        0.238095        0.25000       0.654391   
3        no  0.333333  0.314815        0.238095        0.21875       0.660057   
4        no  0.333333  0.333333        0.190476        0.18750       0.643059   

  winddir16point  weatherCode  \
0             SW      0.01087   
1             SW      0.01087   
2             SW      0.01087   
3             SW      0.01087   
4             SW      0.01087   

                                      weatherIconUrl    weatherDesc  ...  \
0  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
1  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
2  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
3  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
4  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   

   HeatIndexC  HeatIndexF  DewPointC  DewPointF  WindChillC  WindChillF  \
0    0.259259    0.255319     0.5000   0.500000    0.366667    0.370370   
1    0.259259    0.234043     0.5000   0.464286    0.366667    0.351852   
2    0.222222    0.212766     0.4375   0.464286    0.333333    0.333333   
3    0.222222    0.191489     0.4375   0.428571    0.300000    0.314815   
4    0.222222    0.212766     0.4375   0.428571    0.333333    0.333333   

   WindGustMiles  WindGustKmph  FeelsLikeC  FeelsLikeF  
0        0.21875          0.22    0.354839    0.363636  
1        0.21875          0.20    0.354839    0.345455  
2        0.18750          0.18    0.322581    0.327273  
3        0.18750          0.16    0.290323    0.309091  
4        0.15625          0.14    0.322581    0.327273  

[5 rows x 28 columns]

In [41]:
#Total number of unique values in a column
value_counts = []
value_counts_dict = {}
for column in numeric_columns:
  value_counts_dict[column] = len(set(df_min_max_scaled[column]))
  value_counts.append(len(set(df_min_max_scaled[column])))

print(value_counts_dict)

{'tempC': 30, 'tempF': 48, 'windspeedMiles': 22, 'windspeedKmph': 33, 'winddirdegree': 326, 'weatherCode': 15, 'precipMM': 35, 'precipInches': 5, 'humidity': 66, 'visibilityKm': 19, 'visibilityMiles': 12, 'pressureMB': 34, 'pressureInches': 2, 'cloudcover': 101, 'HeatIndexC': 28, 'HeatIndexF': 45, 'DewPointC': 17, 'DewPointF': 29, 'WindChillC': 30, 'WindChillF': 51, 'WindGustMiles': 32, 'WindGustKmph': 47, 'FeelsLikeC': 31, 'FeelsLikeF': 50}


In [42]:
value_sigificance = []
for value in value_counts:
  value_sigificance.append(value/sum(value_counts))

In [43]:
#Minmax normalization of column unique value count
# this will enable us to get the significance of each individual column

import numpy as np

data = np.array(value_sigificance)
normalizedData = (data-np.min(data))/(np.max(data)-np.min(data))
print(normalizedData)

[0.08641975 0.14197531 0.0617284  0.09567901 1.         0.04012346
 0.10185185 0.00925926 0.19753086 0.05246914 0.0308642  0.09876543
 0.         0.30555556 0.08024691 0.13271605 0.0462963  0.08333333
 0.08641975 0.15123457 0.09259259 0.13888889 0.08950617 0.14814815]


In [44]:
df_min_max_scaled

isdaytime     tempC     tempF  windspeedMiles  windspeedKmph  \
0           no  0.366667  0.370370        0.285714        0.28125   
1           no  0.366667  0.351852        0.285714        0.25000   
2           no  0.333333  0.333333        0.238095        0.25000   
3           no  0.333333  0.314815        0.238095        0.21875   
4           no  0.333333  0.333333        0.190476        0.18750   
...        ...       ...       ...             ...            ...   
1459        no  0.500000  0.500000        0.523810        0.53125   
1460        no  0.500000  0.481481        0.523810        0.50000   
1461        no  0.466667  0.462963        0.476190        0.43750   
1462        no  0.433333  0.444444        0.428571        0.43750   
1463        no  0.433333  0.425926        0.380952        0.40625   

      winddirdegree winddir16point  weatherCode  \
0          0.640227             SW     0.010870   
1          0.645892             SW     0.010870   
2          0.654391             SW     0.010870   
3          0.660057             SW     0.010870   
4          0.643059             SW     0.010870   
...             ...            ...          ...   
1459       0.730878              W     0.228261   
1460       0.725212            WSW     0.010870   
1461       0.722380            WSW     0.010870   
1462       0.725212            WSW     0.010870   
1463       0.728045            WSW     0.010870   

                                         weatherIconUrl           weatherDesc  \
0     https://cdn.worldweatheronline.net/images/wsym...         Partly cloudy   
1     https://cdn.worldweatheronline.net/images/wsym...         Partly cloudy   
2     https://cdn.worldweatheronline.net/images/wsym...         Partly cloudy   
3     https://cdn.worldweatheronline.net/images/wsym...         Partly cloudy   
4     https://cdn.worldweatheronline.net/images/wsym...         Partly cloudy   
...                                                 ...                   ...   
1459  https://cdn.worldweatheronline.net/images/wsym...  Patchy rain possible   
1460  https://cdn.worldweatheronline.net/images/wsym...         Partly cloudy   
1461  https://cdn.worldweatheronline.net/images/wsym...         Partly cloudy   
1462  https://cdn.worldweatheronline.net/images/wsym...         Partly cloudy   
1463  https://cdn.worldweatheronline.net/images/wsym...         Partly cloudy   

      ...  HeatIndexC  HeatIndexF  DewPointC  DewPointF  WindChillC  \
0     ...    0.259259    0.255319     0.5000   0.500000    0.366667   
1     ...    0.259259    0.234043     0.5000   0.464286    0.366667   
2     ...    0.222222    0.212766     0.4375   0.464286    0.333333   
3     ...    0.222222    0.191489     0.4375   0.428571    0.300000   
4     ...    0.222222    0.212766     0.4375   0.428571    0.333333   
...   ...         ...         ...        ...        ...         ...   
1459  ...    0.407407    0.404255     0.2500   0.250000    0.500000   
1460  ...    0.407407    0.382979     0.3125   0.285714    0.500000   
1461  ...    0.370370    0.361702     0.3750   0.321429    0.466667   
1462  ...    0.333333    0.340426     0.3750   0.357143    0.433333   
1463  ...    0.333333    0.319149     0.4375   0.428571    0.433333   

      WindChillF  WindGustMiles  WindGustKmph  FeelsLikeC  FeelsLikeF  
0       0.370370        0.21875          0.22    0.354839    0.363636  
1       0.351852        0.21875          0.20    0.354839    0.345455  
2       0.333333        0.18750          0.18    0.322581    0.327273  
3       0.314815        0.18750          0.16    0.290323    0.309091  
4       0.333333        0.15625          0.14    0.322581    0.327273  
...          ...            ...           ...         ...         ...  
1459    0.500000        0.43750          0.44    0.483871    0.490909  
1460    0.481481        0.43750          0.42    0.483871    0.472727  
1461    0.462963        0.40625          0.42    0.451613    0.454545  
1462    0.444444 

In [45]:
column_significance = {}

for i in range(len(normalizedData)):
  column_significance[numeric_columns[i]] = normalizedData[i]

column_significance

{'tempC': 0.08641975308641975,
 'tempF': 0.1419753086419753,
 'windspeedMiles': 0.0617283950617284,
 'windspeedKmph': 0.09567901234567901,
 'winddirdegree': 1.0,
 'weatherCode': 0.04012345679012346,
 'precipMM': 0.10185185185185185,
 'precipInches': 0.009259259259259259,
 'humidity': 0.19753086419753088,
 'visibilityKm': 0.05246913580246914,
 'visibilityMiles': 0.0308641975308642,
 'pressureMB': 0.09876543209876543,
 'pressureInches': 0.0,
 'cloudcover': 0.3055555555555556,
 'HeatIndexC': 0.08024691358024691,
 'HeatIndexF': 0.13271604938271606,
 'DewPointC': 0.046296296296296294,
 'DewPointF': 0.08333333333333333,
 'WindChillC': 0.08641975308641975,
 'WindChillF': 0.15123456790123457,
 'WindGustMiles': 0.09259259259259259,
 'WindGustKmph': 0.1388888888888889,
 'FeelsLikeC': 0.08950617283950617,
 'FeelsLikeF': 0.14814814814814817}

In [46]:
def get_entry_significance(parent_col, entry):
  if(str(entry) == 'nan'):
    return 0
  all_entries = len(df_min_max_scaled[parent_col])
  items = df_min_max_scaled[parent_col].value_counts()[entry]
  return items/all_entries


In [47]:
#GET sigificace of each etry in the 'table'
#We repeat the same techniques we used for the columns
# entry_significance = {}
# entry_significance_list = []
# for column in numeric_columns:
#   items = df_min_max_scaled[column]
#   respective = []
#   for item in items:
#     respective.append(get_entry_significance(column,item))
#   entry_significance_list.append(respective)
#   entry_significance[column] = respective

In [48]:
# for column in numeric_columns:
#   df_min_max_scaled[column] = entry_significance[column]

In [49]:
df_min_max_scaled.head()
#Unscaled significance values

isdaytime     tempC     tempF  windspeedMiles  windspeedKmph  winddirdegree  \
0        no  0.366667  0.370370        0.285714        0.28125       0.640227   
1        no  0.366667  0.351852        0.285714        0.25000       0.645892   
2        no  0.333333  0.333333        0.238095        0.25000       0.654391   
3        no  0.333333  0.314815        0.238095        0.21875       0.660057   
4        no  0.333333  0.333333        0.190476        0.18750       0.643059   

  winddir16point  weatherCode  \
0             SW      0.01087   
1             SW      0.01087   
2             SW      0.01087   
3             SW      0.01087   
4             SW      0.01087   

                                      weatherIconUrl    weatherDesc  ...  \
0  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
1  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
2  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
3  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
4  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   

   HeatIndexC  HeatIndexF  DewPointC  DewPointF  WindChillC  WindChillF  \
0    0.259259    0.255319     0.5000   0.500000    0.366667    0.370370   
1    0.259259    0.234043     0.5000   0.464286    0.366667    0.351852   
2    0.222222    0.212766     0.4375   0.464286    0.333333    0.333333   
3    0.222222    0.191489     0.4375   0.428571    0.300000    0.314815   
4    0.222222    0.212766     0.4375   0.428571    0.333333    0.333333   

   WindGustMiles  WindGustKmph  FeelsLikeC  FeelsLikeF  
0        0.21875          0.22    0.354839    0.363636  
1        0.21875          0.20    0.354839    0.345455  
2        0.18750          0.18    0.322581    0.327273  
3        0.18750          0.16    0.290323    0.309091  
4        0.15625          0.14    0.322581    0.327273  

[5 rows x 28 columns]

In [50]:
#Normalize values to be between 0 and 1
for numeric_column in numeric_columns:
  df_min_max_scaled[numeric_column] = (df_min_max_scaled[numeric_column] - df_min_max_scaled[numeric_column].min()) / (df_min_max_scaled[numeric_column].max() - df_min_max_scaled[numeric_column].min())

In [51]:
#now we have a table populated with significance values that have een scaled to be between 0 and 1
df_min_max_scaled.head()

isdaytime     tempC     tempF  windspeedMiles  windspeedKmph  winddirdegree  \
0        no  0.366667  0.370370        0.285714        0.28125       0.640227   
1        no  0.366667  0.351852        0.285714        0.25000       0.645892   
2        no  0.333333  0.333333        0.238095        0.25000       0.654391   
3        no  0.333333  0.314815        0.238095        0.21875       0.660057   
4        no  0.333333  0.333333        0.190476        0.18750       0.643059   

  winddir16point  weatherCode  \
0             SW      0.01087   
1             SW      0.01087   
2             SW      0.01087   
3             SW      0.01087   
4             SW      0.01087   

                                      weatherIconUrl    weatherDesc  ...  \
0  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
1  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
2  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
3  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
4  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   

   HeatIndexC  HeatIndexF  DewPointC  DewPointF  WindChillC  WindChillF  \
0    0.259259    0.255319     0.5000   0.500000    0.366667    0.370370   
1    0.259259    0.234043     0.5000   0.464286    0.366667    0.351852   
2    0.222222    0.212766     0.4375   0.464286    0.333333    0.333333   
3    0.222222    0.191489     0.4375   0.428571    0.300000    0.314815   
4    0.222222    0.212766     0.4375   0.428571    0.333333    0.333333   

   WindGustMiles  WindGustKmph  FeelsLikeC  FeelsLikeF  
0        0.21875          0.22    0.354839    0.363636  
1        0.21875          0.20    0.354839    0.345455  
2        0.18750          0.18    0.322581    0.327273  
3        0.18750          0.16    0.290323    0.309091  
4        0.15625          0.14    0.322581    0.327273  

[5 rows x 28 columns]

In [52]:
# df_min_max_scaled = df_min_max_scaled.drop(["date", "time"], axis=1)

In [53]:
# #Adding the column significance to the row significance

# for column in numeric_columns:
#   df_min_max_scaled[column] = df_min_max_scaled[column] * column_significance[column]

In [54]:
df_min_max_scaled.head(20)

isdaytime     tempC     tempF  windspeedMiles  windspeedKmph  \
0         no  0.366667  0.370370        0.285714        0.28125   
1         no  0.366667  0.351852        0.285714        0.25000   
2         no  0.333333  0.333333        0.238095        0.25000   
3         no  0.333333  0.314815        0.238095        0.21875   
4         no  0.333333  0.333333        0.190476        0.18750   
5         no  0.366667  0.370370        0.190476        0.15625   
6         no  0.400000  0.388889        0.190476        0.15625   
7         no  0.433333  0.425926        0.190476        0.15625   
8        yes  0.466667  0.481481        0.190476        0.15625   
9        yes  0.533333  0.518519        0.190476        0.18750   
10       yes  0.566667  0.574074        0.238095        0.25000   
11       yes  0.633333  0.629630        0.285714        0.28125   
12       yes  0.666667  0.685185        0.333333        0.34375   
13       yes  0.666667  0.685185        0.380952        0.37500   
14       yes  0.700000  0.685185        0.428571        0.40625   
15       yes  0.700000  0.685185        0.428571        0.43750   
16       yes  0.666667  0.666667        0.428571        0.40625   
17        no  0.666667  0.666667        0.428571        0.40625   
18        no  0.633333  0.648148        0.428571        0.40625   
19        no  0.600000  0.611111        0.380952        0.37500   

    winddirdegree winddir16point  weatherCode  \
0        0.640227             SW      0.01087   
1        0.645892             SW      0.01087   
2        0.654391             SW      0.01087   
3        0.660057             SW      0.01087   
4        0.643059             SW      0.01087   
5        0.628895             SW      0.01087   
6        0.611898             SW      0.01087   
7        0.600567             SW      0.01087   
8        0.592068            SSW      0.01087   
9        0.580737            SSW      0.01087   
10       0.572238            SSW      0.01087   
11       0.563739            SSW      0.01087   
12       0.558074            SSW      0.01087   
13       0.560907            SSW      0.01087   
14       0.563739            SSW      0.01087   
15       0.566572            SSW      0.01087   
16       0.555241            SSW      0.01087   
17       0.546742            SSW      0.01087   
18       0.538244            SSW      0.01087   
19       0.515581              S      0.01087   

                                       weatherIconUrl    weatherDesc  ...  \
0   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
1   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
2   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
3   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
4   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
5   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
6   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
7   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
8   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
9   https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
10  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
11  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
12  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
13  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
14  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
15  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
16  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
17  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
18  https://cdn.worldweatheronline.net/images/wsym...  Partly cloudy  ...   
19  https://cdn.worldweat

In [55]:
df_num = df_min_max_scaled[numeric_columns]
# df_num.drop(["loc_id"], axis=1)

In [56]:
#Creates a list of all non numeric colums
non_numeric = set(df.columns).difference(set(numeric_columns))

In [57]:
df_non_numeric = df[list(non_numeric)]

In [58]:
df_non_numeric.head()

weatherIconUrl winddir16point  \
0  https://cdn.worldweatheronline.net/images/wsym...             SW   
1  https://cdn.worldweatheronline.net/images/wsym...             SW   
2  https://cdn.worldweatheronline.net/images/wsym...             SW   
3  https://cdn.worldweatheronline.net/images/wsym...             SW   
4  https://cdn.worldweatheronline.net/images/wsym...             SW   

     weatherDesc isdaytime  
0  Partly cloudy        no  
1  Partly cloudy        no  
2  Partly cloudy        no  
3  Partly cloudy        no  
4  Partly cloudy        no

In [59]:
df_num.head()

tempC     tempF  windspeedMiles  windspeedKmph  winddirdegree  \
0  0.366667  0.370370        0.285714        0.28125       0.640227   
1  0.366667  0.351852        0.285714        0.25000       0.645892   
2  0.333333  0.333333        0.238095        0.25000       0.654391   
3  0.333333  0.314815        0.238095        0.21875       0.660057   
4  0.333333  0.333333        0.190476        0.18750       0.643059   

   weatherCode  precipMM  precipInches  humidity  visibilityKm  ...  \
0      0.01087       0.0           0.0  0.772727      0.444444  ...   
1      0.01087       0.0           0.0  0.787879      0.444444  ...   
2      0.01087       0.0           0.0  0.803030      0.444444  ...   
3      0.01087       0.0           0.0  0.803030      0.444444  ...   
4      0.01087       0.0           0.0  0.772727      0.444444  ...   

   HeatIndexC  HeatIndexF  DewPointC  DewPointF  WindChillC  WindChillF  \
0    0.259259    0.255319     0.5000   0.500000    0.366667    0.370370   
1    0.259259    0.234043     0.5000   0.464286    0.366667    0.351852   
2    0.222222    0.212766     0.4375   0.464286    0.333333    0.333333   
3    0.222222    0.191489     0.4375   0.428571    0.300000    0.314815   
4    0.222222    0.212766     0.4375   0.428571    0.333333    0.333333   

   WindGustMiles  WindGustKmph  FeelsLikeC  FeelsLikeF  
0        0.21875          0.22    0.354839    0.363636  
1        0.21875          0.20    0.354839    0.345455  
2        0.18750          0.18    0.322581    0.327273  
3        0.18750          0.16    0.290323    0.309091  
4        0.15625          0.14    0.322581    0.327273  

[5 rows x 24 columns]

In [60]:
import torch
row_attention_matrices = []

for column in df_num.columns:
  Table_1 = []
  for column_2 in df_num.columns:
    c = df_num[column] * df_num[column_2]
    Table_1.append(torch.tensor(list(c)))
  row_attention_matrices.append(torch.stack(Table_1))
x = torch.stack(row_attention_matrices)
x_T = torch.transpose(x,2,0)
print(f"GENERAL SHAPE: {x_T.shape}")
print(f"SHAPE OF A SINGLE LAYER: {x_T[6].shape}")

ModuleNotFoundError: No module named 'torch'

A major problem we face with our data is the fact that most of our data is numeric and not text. All the systems that summarization system used, are specifically made for text data. This means that we have somehow have to make our numeric data behave like text so that we can summarize it. Upto this point, we have been working to make our numeric data behave like text data. And now we have finally reached the last part of the process. From the above output, you can see that there are 61  30x30 matrices. Each 30x30 matrix is a result of us calculating the signnificance of our numeric columns against themselves. 1464 is the number of rows. These 30X30 matrices are important since they will help us get the embeddings for our numeric data as we pass them through an RNN model.

In [ ]:
all = df_num.values.tolist()
all_test = all[int(len(all) * 0.8):]
all = all[:int(len(all) * 0.8)]
print(all[0])

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
import numpy as np
from numpy.linalg import norm

def COS_SIM(A,B):
  A = np.array(A)
  B = np.array(B)
  cosine = np.dot(A,B)/(norm(A)*norm(B))
  return cosine


In [ ]:
'''Preprocessing the text part of our data'''

In [ ]:
len(all)

In [ ]:
#Non numeric data
df_non_numeric.head()

In [ ]:
df_non_numeric = df_non_numeric.drop(["weatherIconUrl"], axis=1)
df_non_numeric
#We droped the icon url since we do not need it for out use case

In [ ]:
#Get a list of words
all_words = []
for column in df_non_numeric.columns:
  column_words = set(df_non_numeric[column])
  all_words.extend(list(column_words))

print(all_words)

In [ ]:
#Give each entry a number that will represent it
WORD_2_TOKEN = {}
TOKEN_2_WORD = {}

def Tokenize(word, i):
  WORD_2_TOKEN[word] = i
  TOKEN_2_WORD[i] = word
  return

def encode(row):
  encoded = []
  for word in row:
    encoded.append(WORD_2_TOKEN[word])
  return encoded

def decode(row_encoded):
  decoded = []
  for word in row_encoded:
    decoded.append(TOKEN_2_WORD[word])
  return decoded


In [ ]:
for i in range(len(all_words)):
  Tokenize(all_words[i], i)

In [ ]:
print(f"Encode example: {encode(['Light rain shower','Partly cloudy'])}")

In [ ]:
print(f"Decode example: {decode([29,19])}")

Now our text data is ready to be passed to a deep learning model.

During training, we will pass the numerical representation for our word through an embedding layer which will produce word embeddings. As for the part concerning numeric data, we will take our 26X26 matrix and put it through a softmax layer and then through a linear layer to obtain the embeddings for our


In [ ]:
import torch
def tensorfy(chosen_sequence):
  csq = []
  for x_seq in chosen_sequence:
    csq.append(torch.tensor(x_seq))
  return csq

In [ ]:
from random import randint

def get_sequence(SEQ_LEN):
  start_point = randint(0,( len(all) - SEQ_LEN) -1)
  chosen_sequence = all[start_point: start_point + SEQ_LEN]
  return tensorfy(chosen_sequence),chosen_sequence

def get_test_sequence(SEQ_LEN):
  start_point = randint(0,( len(all_test) - SEQ_LEN) -1)
  chosen_sequence = all_test[start_point: start_point + SEQ_LEN]
  return tensorfy(chosen_sequence),chosen_sequence

In [ ]:
def get_max_index(src_input):
  local = src_input.copy()
  local.sort()
  return local[len(local) - 2]

In [ ]:
def get_maxes(cos_sims):
  maxes = []
  for cos_sim in cos_sims:
    maxes.append(torch.tensor(get_max_index(cos_sim)))
  return torch.stack(maxes)

In [ ]:
def get_batch(Batch_Size,chunk="train"):
  Batch_Size = 32
  batch = []
  labels = []
  vanilla_batch = []
  for _ in range(Batch_Size):
    if(chunk == "train"):
      src_input, vnl_input = get_sequence(64)
    else:
      src_input, vnl_input = get_test_sequence(64)
    # labels.append(torch.tensor(get_max_index(src_input)))
    src_input = torch.stack(src_input).transpose(1,0)
    cos_sims = cosine_similarity(detensorfy(src_input))
    labels.append(get_maxes(cos_sims))
    batch.append(src_input)
    vanilla_batch.append(vnl_input)
  return torch.stack(batch),torch.stack(labels)

In [ ]:
def detensorfy(tensor_array):
  arr = []
  for tnsr in tensor_array:
    arr.append(tnsr.tolist())
  return arr

In [ ]:
import math
import torch.nn.functional as F

# Given Query, Key, Value, calculate the final weighted value
def scaled_dot_product_attention(q, k, v, mask=None, dropout=None):
    attention_scores = torch.matmul(q, k.transpose(-2, -1))/math.sqrt(q.shape[-1])
    if mask is not None:
        attention_scores = attention_scores.masked_fill(mask == 0, value=-1e9)

    # Softmax along the last dimension
    attention_weights = F.softmax(attention_scores, dim=-1)

    if dropout is not None:
        attention_weights = dropout(attention_weights)

    output = torch.matmul(attention_weights, v)
    return output

In [ ]:
class MultiHeadAttention(torch.nn.Module):
    def __init__(self, n_heads, d_model, dropout=0.1):
        super().__init__()

        self.n_heads = n_heads
        self.d_model = d_model
        self.d_k = self.d_k = self.d_v = d_model//n_heads

        # self attention linear layers
        # Linear layers for q, k, v vectors generation in different heads
        self.q_linear_layers = []
        self.k_linear_layers = []
        self.v_linear_layers = []
        for i in range(n_heads):

            self.q_linear_layers.append(torch.nn.Linear(d_model, self.d_k).to(device))
            self.k_linear_layers.append(torch.nn.Linear(d_model, self.d_k).to(device))
            self.v_linear_layers.append(torch.nn.Linear(d_model, self.d_v).to(device))

        self.dropout = torch.nn.Dropout(dropout)
        self.out = torch.nn.Linear(n_heads*self.d_v, d_model)

    def forward(self, q, k, v, mask=None):
        multi_head_attention_outputs = []
        for q_linear, k_linear, v_linear in zip(self.q_linear_layers,
                                                self.k_linear_layers,
                                                self.v_linear_layers):
            new_q = q_linear(q)
            new_k = k_linear(k)
            new_v = v_linear(v)


            head_v = scaled_dot_product_attention(new_q, new_k, new_v, mask, self.dropout)  # (batch_size, seq_len, d_v)
            multi_head_attention_outputs.append(head_v)

        concat = torch.cat(multi_head_attention_outputs, -1)  # (batch_size, seq_len, n_heads*d_v)

        # Linear layer to recover to original shap
        output = self.out(concat)  # (batch_size, seq_len, d_model)

        return output


In [ ]:
class FeedForward(torch.nn.Module):
    def __init__(self, d_model, d_ff=2048, dropout=0.1):
        super().__init__()

        self.linear_1 = torch.nn.Linear(d_model, d_ff)
        self.dropout = torch.nn.Dropout(dropout)
        self.linear_2 = torch.nn.Linear(d_ff, d_model)

    def forward(self, x):
        x = self.dropout(F.relu(self.linear_1(x)))
        x = self.linear_2(x)
        return x

In [ ]:
class LayerNorm(torch.nn.Module):
    def __init__(self, d_model, eps=1e-6):
        super().__init__()
        self.d_model = d_model
        self.alpha = torch.nn.Parameter(torch.ones(self.d_model))
        self.beta = torch.nn.Parameter(torch.zeros(self.d_model))
        self.eps = eps

    def forward(self, x):
        # x size: (batch_size, seq_len, d_model)
        x_hat = (x - x.mean(dim=-1, keepdim=True))/(x.std(dim=-1, keepdim=True) + self.eps)
        x_tilde = self.alpha*x_hat + self.beta
        return x_tilde

In [ ]:
class EncoderLayer(torch.nn.Module):
    def __init__(self, d_model, n_heads, dropout=0.1):
        super().__init__()
        self.d_model = d_model
        self.norm_1 = LayerNorm(d_model)
        self.norm_2 = LayerNorm(d_model)
        self.multi_head_attention = MultiHeadAttention(n_heads, d_model).to(device)
        self.feed_forward = FeedForward(d_model)
        self.dropout_1 = torch.nn.Dropout(dropout).to(device)
        self.dropout_2 = torch.nn.Dropout(dropout).to(device)

    def forward(self, x, mask):
        #import pdb; pdb.set_trace()

        x = x + self.dropout_1(self.multi_head_attention(x, x, x, mask).to(device)).to(device)
        x = self.norm_1(x)
        x = x + self.dropout_2(self.feed_forward(x))
        x = self.norm_2(x)
        return x

In [ ]:
class MLP(torch.nn.Module):
  def __init__(self,d_model):
    super().__init__()
    self.linear = torch.nn.Linear(d_model,1024)
    self.linear_2 = torch.nn.Linear(1024,24)

  def forward(self, x):
    x = self.linear(x)
    x = F.relu(x)
    x = self.linear_2(x)
    return x

In [ ]:
import copy

def clone_layer(module, N):
    return torch.nn.ModuleList([copy.deepcopy(module) for i in range(N)])

In [ ]:
class Encoder(torch.nn.Module):
    def __init__(self, d_model, N, n_heads):
        super().__init__()
        self.encoder_layers = clone_layer(EncoderLayer(d_model, n_heads).to(device), N).to(device)
        self.norm = LayerNorm(d_model).to(device)
        self.MLP = MLP(64 * 24)
    def forward(self, src_embeddings, mask):
        x = src_embeddings
        for encoder in self.encoder_layers:
            x = encoder(x, mask)
        x = self.norm(x)
        x = torch.flatten(x, 1,2)
        x = self.MLP(x)
        x = F.tanh(x)
        return x

In [ ]:
d_model = 64
n_heads = 2
N = 1

In [ ]:
device = "cuda"

In [ ]:
Train_model = Encoder(d_model, N, n_heads).to(device)

In [ ]:
for p in Train_model.parameters():
    if p.dim() > 1:
        torch.nn.init.xavier_uniform(p)

In [ ]:
optimizer = torch.optim.Adam(Train_model.parameters(),lr=1e-5)

In [ ]:
device = "cuda"

In [ ]:
criterion = torch.nn.MSELoss()

In [ ]:
train_losses = []
epochs = 100
Train_model.train()
i = 0
for e in range(epochs):
  for _ in range(40):
    batch, labels = get_batch(32, chunk="train")
    preds = Train_model(batch.to(device),None)
    optimizer.zero_grad()
    loss = criterion(preds, labels.to(device).float())
    train_losses.append(loss.item())
    loss.backward()
    optimizer.step()
    i += 1
    if(i % 10 == 0):
      print(f"[-] Train_loss is {np.mean(train_losses)} iteration is {i} on epoch {e}")
      train_losses = []
  i = 0


In [ ]:
from sklearn.metrics import mean_squared_error
def get_mse_values(y_preds, y_true):
  mse_losses = []
  i = 0
  for pred in y_preds:
    mse_losses.append(mean_squared_error(y_true[i].tolist(),pred.tolist()))
    i += 1
  return torch.tensor(mse_losses)

def get_rmse_values(y_preds, y_true):
  mse_losses = []
  i = 0
  for pred in y_preds:
    mse_losses.append(mean_squared_error(y_true[i].tolist(),pred.tolist(), squared=False))
    i += 1
  return torch.tensor(mse_losses)

In [ ]:
train_losses = []
i = 0
MSE_LOSSES = []
All_losses = []
RMSE_losses = []
for _ in range(100):
  batch, labels = get_batch(32, chunk="test")
  preds = Train_model(batch.to(device),None)
  optimizer.zero_grad()
  loss = criterion(preds, labels.to(device).float())
  train_losses.append(loss.item())
  preds_transposed = preds.transpose(1,0)
  labels_transposed = labels.transpose(1,0)
  All_losses.append(get_mse_values(preds_transposed, labels_transposed))
  RMSE_losses.append(get_rmse_values(preds_transposed, labels_transposed))
  MSE_LOSSES.append(torch.cosine_similarity(preds_transposed,labels_transposed.to(device)))
  i += 1
  if(i % 5 == 0):
    print(f"[-] Test_loss is {np.mean(train_losses)} iteration is {i}")
    train_losses = []
i = 0

In [ ]:
Cos_sim_stacked = torch.stack(MSE_LOSSES).mean(0)
i = 1
print(f"Class Class_Name Cosine Similarities \n")
for loss in Cos_sim_stacked.tolist():
  print(f"{i} {numeric_columns[i-1]} {loss}")
  i += 1

In [ ]:
from matplotlib import pyplot as plt
classes = [i for i in range(1,25)]
fig = plt.figure(figsize =(10, 5))
plt.title("Prediction vs Labels Cosine Similarities")
plt.bar(classes, Cos_sim_stacked.tolist())
plt.show()

In [ ]:
All_losses_stacked = torch.stack(All_losses).mean(0)
from matplotlib import pyplot as plt
classes = [i for i in range(1,25)]
fig = plt.figure(figsize =(5, 3))
plt.title("Transformer Mean Squared Error")
plt.bar(classes, All_losses_stacked.tolist())
plt.show()

In [ ]:
print(f"Class Class_Name Mean Squared Error \n")
i = 1
for loss in All_losses_stacked.tolist():
  print(f"{i} {numeric_columns[i-1]} {loss}")
  i += 1

In [ ]:
RMSE_losses_stacked = torch.stack(RMSE_losses).mean(0)
i = 1
print(f"Class RMSE Error \n")

for loss in RMSE_losses_stacked.tolist():
  print(f"{i} {loss}")
  i += 1


In [ ]:

from matplotlib import pyplot as plt
classes = [i for i in range(1,25)]
fig = plt.figure(figsize =(5, 3))
plt.title("Transformer Root Mean Squared Error(RMSE)")
plt.bar(classes, RMSE_losses_stacked.tolist())
plt.show()

In [ ]:
from torch import nn
class DetectorRNN(nn.Module):
    def __init__(self,no_layers,hidden_dim,embedding_dim,drop_prob=0.5):
        super(DetectorRNN,self).__init__()

        self.output_dim = output_dim
        self.hidden_dim = hidden_dim

        self.no_layers = no_layers
        # self.vocab_size = vocab_size

        # self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(input_size=embedding_dim,hidden_size=self.hidden_dim,
                           num_layers=no_layers, batch_first=True)


        # dropout layer
        self.dropout = nn.Dropout(0.3)

        # linear and sigmoid layer
        self.fc = nn.Linear(self.hidden_dim, output_dim)
        # self.sig = nn.Softmax(dim=)

    def forward(self,x,hidden):
        batch_size = x.size(0)

        lstm_out, hidden = self.lstm(x, hidden)

        out = self.dropout(lstm_out)
        sig_out = self.fc(out)


        sig_out = sig_out[:, -1] # get last batch of labels

        return sig_out, hidden

    def init_hidden(self, batch_size):
        h0 = torch.zeros((self.no_layers,batch_size,self.hidden_dim)).to(device)
        c0 = torch.zeros((self.no_layers,batch_size,self.hidden_dim)).to(device)
        hidden = (h0,c0)
        return hidden

In [ ]:
no_layers = 1
embedding_dim = 64
output_dim = 24
hidden_dim = 200

In [ ]:
Detector_model = DetectorRNN(1,hidden_dim,embedding_dim,drop_prob=0.1).to(device)

In [ ]:
optimizer = torch.optim.Adam(Detector_model.parameters(),lr=1e-5)

In [ ]:

epoch = 100
Detector_model.train()
Total_loss = 0
h = Detector_model.init_hidden(32)
clip = 5

for e in range(epoch):
  i = 0
  for _ in range(40):
    batch, labels = get_batch(32, chunk="train")
    h = tuple([each.data for each in h])
    Detector_model.zero_grad()
    preds,h = Detector_model(batch.to(device),h)
    loss = criterion(preds, labels.to(device).float())
    Total_loss += loss.item()
    loss.backward()
    nn.utils.clip_grad_norm_(Detector_model.parameters(), clip)
    optimizer.step()
    i += 1
    if(i % 20 == 0):
      print(f"[~] Epoch = {e+1} itteration = {i} Train_loss = {Total_loss/20}")
      Total_loss = 0


In [ ]:
i = 0
Total_loss = 0
MSE_LOSSES = []
All_losses = []
RMSE_losses = []
for _ in range(100):
  batch, labels = get_batch(32, chunk="test")
  h = tuple([each.data for each in h])
  Detector_model.zero_grad()
  preds,h = Detector_model(batch.to(device),h)
  loss = criterion(preds, labels.to(device).float())
  Total_loss += loss.item()
  preds_transposed = preds.transpose(1,0)
  labels_transposed = labels.transpose(1,0)
  All_losses.append(get_mse_values(preds_transposed, labels_transposed))
  RMSE_losses.append(get_rmse_values(preds_transposed, labels_transposed))
  MSE_LOSSES.append(torch.cosine_similarity(preds_transposed,labels_transposed.to(device)))
  i += 1
  if(i % 20 == 0):
    print(f"[~] Epoch = {e+1} itteration = {i} Train_loss = {Total_loss/20}")
    Total_loss = 0

In [ ]:
Cos_sim_stacked = torch.stack(MSE_LOSSES).mean(0)
i = 1
print(f"Class Class_Name Cosine Similarities \n")
for loss in Cos_sim_stacked.tolist():
  print(f"{i} {numeric_columns[i-1]} {loss}")
  i += 1

In [ ]:
from matplotlib import pyplot as plt
classes = [i for i in range(1,25)]
fig = plt.figure(figsize =(10, 5))
plt.title("LSTM Predictions vs Labels Cosine Similarities")
plt.bar(classes, Cos_sim_stacked.tolist())
plt.show()

In [ ]:
All_losses_stacked = torch.stack(All_losses).mean(0)
from matplotlib import pyplot as plt
classes = [i for i in range(1,25)]
fig = plt.figure(figsize =(5, 3))
plt.title("LSTM Mean Squared Error")
plt.bar(classes, All_losses_stacked.tolist())
plt.show()

In [ ]:
print(f"Class Class_Name Mean Squared Error \n")
i = 1
for loss in All_losses_stacked.tolist():
  print(f"{i} {numeric_columns[i-1]} {loss}")
  i += 1

In [ ]:
RMSE_losses_stacked = torch.stack(RMSE_losses).mean(0)
i = 1
print(f"Class RMSE Error \n")

for loss in RMSE_losses_stacked.tolist():
  print(f"{i} {loss}")
  i += 1

In [ ]:
from matplotlib import pyplot as plt
classes = [i for i in range(1,25)]
fig = plt.figure(figsize =(5, 3))
plt.title("LSTM Root Mean Squared Error(RMSE)")
plt.bar(classes, RMSE_losses_stacked.tolist())
plt.show()